src/plugins/controls/public/options_list/embeddable/options_list_embeddable.tsx

```diff
@@ -207,30 +207,31 @@ export class OptionsListEmbeddable
               a.exclude === b.exclude &&
               a.existsSelected === b.existsSelected &&
               isEqual(a.selectedOptions, b.selectedOptions)
-          )
-        )
-        .subscribe(async ({ selectedOptions: newSelectedOptions }) => {
-          if (!newSelectedOptions || isEmpty(newSelectedOptions)) {
-            this.dispatch.clearValidAndInvalidSelections({});
-          } else {
-            const { invalidSelections } = this.getState().componentState ?? {};
-            const newValidSelections: string[] = [];
-            const newInvalidSelections: string[] = [];
-            for (const selectedOption of newSelectedOptions) {
-              if (invalidSelections?.includes(selectedOption)) {
-                newInvalidSelections.push(selectedOption);
-                continue;
+          ),
+          mergeMap(async ({ selectedOptions: newSelectedOptions }) => {
```

Comment:
It seems like a `switchMap` might be a better option here. If the selected options change while the filter is being built, I would expect the previous observable to be canceled. Additionally, we could move the synchronous logic out into a `tap` instead so it runs every time. I tested it with the following code: and it worked as expected.

```
          tap(({ selectedOptions: newSelectedOptions }) => {
            if (!newSelectedOptions || isEmpty(newSelectedOptions)) {
              this.dispatch.clearValidAndInvalidSelections({});
            } else {
              const { invalidSelections } = this.getState().componentState ?? {};
              const newValidSelections: string[] = [];
              const newInvalidSelections: string[] = [];
              for (const selectedOption of newSelectedOptions) {
                if (invalidSelections?.includes(selectedOption)) {
                  newInvalidSelections.push(selectedOption);
                  continue;
                }
                newValidSelections.push(selectedOption);
              }
              this.dispatch.setValidAndInvalidSelections({
                validSelections: newValidSelections,
                invalidSelections: newInvalidSelections,
              });
            }
          }),
          switchMap(async () => {
            const newFilters = await this.buildFilter();
            this.dispatch.publishFilters(newFilters);
          })
```

Reply:
Addressed in [`0fab35e` (#169522)](https://github.com/elastic/kibana/pull/169522/commits/0fab35e43a6392ec41e358d0cd95140323fac58a)